src/mapping/query/function/flatten.rs

```diff
@@ -0,0 +1,362 @@
+                    self.new_key(key),
+                    value.iter(),
+                )));
+                self.next()
+            }
+            Some((key, value)) => Some((self.new_key(key), value)),
+            None => None,
+        }
+    }
+}
+
+/// Create an iterator that can walk a tree of Array values.
+/// This can be used to flatten the array.
+struct ValueFlatten<'a> {
+    values: std::slice::Iter<'a, Value>,
+    inner: Option<Box<ValueFlatten<'a>>>,
+}
+
+impl<'a> ValueFlatten<'a> {
+    fn new(values: std::slice::Iter<'a, Value>) -> Self {
+        ValueFlatten {
+            values,
+            inner: None,
+        }
+    }
+}
+
+impl<'a> std::iter::Iterator for ValueFlatten<'a> {
+    type Item = &'a Value;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // Iterate over our inner list first.
+        if let Some(ref mut inner) = self.inner {
+            let next = inner.next();
+            if next.is_some() {
+                return next;
+            } else {
+                // The inner list has been exhausted.
+                self.inner = None;
+            }
+        }
+
+        // Then iterate over our values.
+        let next = self.values.next();
+        if let Some(Value::Array(next)) = next {
+            // Create a new iterator for this child list.
+            self.inner = Some(Box::new(ValueFlatten::new(next.iter())));
+            self.next()
+        } else {
+            next
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) struct FlattenFn {
+    value: Box<dyn Function>,
+}
+
+impl FlattenFn {
+    #[cfg(test)]
+    pub(in crate::mapping) fn new(value: Box<dyn Function>) -> Self {
+        FlattenFn { value }
+    }
+}
+
+impl Function for FlattenFn {
+    fn execute(&self, ctx: &Event) -> Result<QueryValue> {
+        let value = required_value!(ctx, self.value,
+        Value::Array(arr) => Value::Array(
+            ValueFlatten::new(arr.iter()).cloned().collect()
+        ),
+        Value::Map(map) => Value::Map(
+            MapFlatten::new(map.iter()).map(|(k, v)| (k, v.clone())).collect()
+        ));
+
+        Ok(value.into())
+    }
+
+    fn parameters() -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| {
+                matches!(v, QueryValue::Value(Value::Array(_)) |
+                                      QueryValue::Value(Value::Map(_))
+                )
+            },
+            required: true,
+        }]
+    }
+}
+
+impl TryFrom<ArgumentList> for FlattenFn {
+    type Error = String;
+
+    fn try_from(mut arguments: ArgumentList) -> Result<Self> {
+        let value = arguments.required("value")?;
+        Ok(Self { value })
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::mapping::query::path::Path;
+    use serde_json::json;
+
+    #[test]
+    fn check_flatten() {
+        let cases = vec![
```

Comment:
I believe these cases will become quite a bit more readable by applying new testing patterns, such as can be seen in this PR:

https://github.com/timberio/vector/pull/4770

Basically;

* use `map![]` to create maps
* use `into()` vs the more verbose `Value::from`
* don't test against an actual event, but use a map (`std::collections::BTreeMap` implements `remap::Object`)
* paths can be created from strings, so change `Path::from(vec![vec!["foo"]])` to `Path::from("foo")`
* test against an actual path only once, use `Literal::from(map!["foo": "bar"])` for other tests (and leave the event an empty `map![]`), to reduce the amount of unneeded boilerplate.

I realize this requires the PR to be rebased on the new remap-lang work first, so I propose we keep things as they are in this PR, and then refactor once the remap-lang PRs are merged, at which point we'll have to change this implementation anyway.